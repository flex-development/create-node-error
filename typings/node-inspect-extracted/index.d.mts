declare module 'node-inspect-extracted' {
  import type { ObjectPlain } from '@flex-development/tutils'

  export interface InspectOptions {
    breakLength?: number | undefined
    colors?: boolean | undefined
    compact?: boolean | number | undefined
    customInspect?: boolean | undefined
    depth?: number | null | undefined
    getters?: boolean | 'get' | 'set' | undefined
    maxArrayLength?: number | null | undefined
    maxStringLength?: number | null | undefined
    numericSeparator?: boolean | undefined
    showHidden?: boolean | undefined
    showProxy?: boolean | undefined
    sorted?: boolean | ((a: string, b: string) => number) | undefined
  }

  export interface InspectOptionsStylized extends InspectOptions {
    stylize(text: string, styleType: Style): string
  }

  export type Style =
    | 'bigint'
    | 'boolean'
    | 'date'
    | 'module'
    | 'null'
    | 'number'
    | 'regexp'
    | 'special'
    | 'string'
    | 'symbol'
    | 'undefined'

  export type CustomInspectFunction = (
    depth: number,
    options: InspectOptionsStylized
  ) => string

  export function format(format?: any, ...param: any[]): string

  export function formatWithOptions(
    inspectOptions: InspectOptions,
    format?: any,
    ...param: any[]
  ): string

  export function inspect(
    object: any,
    showHidden?: boolean,
    depth?: number | null,
    color?: boolean
  ): string

  export function inspect(object: any, options?: InspectOptions): string

  export namespace inspect {
    let colors: Record<number, number>
    let styles: { [K in Style]: string }
    let defaultOptions: InspectOptions
    let replDefaults: InspectOptions
    const custom: unique symbol
  }

  export function stripVTControlCharacters(str: string): string

  export function stylizeWithColor(text: string, styleType: Style): string

  export function stylizeWithHTML(text: string, styleType: Style): string

  export const Proxy: new <T extends ObjectPlain>(
    target: T,
    handler: ProxyHandler<T>
  ) => T
}
