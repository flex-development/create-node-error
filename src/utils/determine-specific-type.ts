/**
 * @file Utilities - determineSpecificType
 * @module errnode/utils/determineSpecificType
 * @see https://github.com/nodejs/node/blob/v22.7.0/lib/internal/errors.js#L988-L1041
 */

import { inspect } from 'node-inspect-extracted'

/**
 * Determine the specific type of a value for type-mismatch errors.
 *
 * @param {unknown} value - Value to check
 * @return {string} Specific type of `value`
 */
function determineSpecificType(value: unknown): string {
  /**
   * Type of {@linkcode value}.
   *
   * @const {string} type
   */
  const type: string = typeof value

  /**
   * Specific type.
   *
   * @var {string} specificType
   */
  let specificType: string = ''

  switch (type) {
    case 'object':
    case 'undefined':
      specificType = value?.constructor && 'name' in value.constructor
        ? `an instance of ${value.constructor.name}`
        : inspect(value, { depth: -1 })
      break
    case 'function':
      specificType = `type ${type} ${(<(...args: any[]) => any>value).name}`
      break
    case 'string':
      // @ts-expect-error `value` is a string (18046).
      value.length > 28 && (value = `${(<string>value).slice(0, 25)}...`)
      specificType = `type ${type} (${inspect(value)})`
      break
    default:
      specificType = `type ${type} (${inspect(value)})`
      break
  }

  return specificType
}

export default determineSpecificType
