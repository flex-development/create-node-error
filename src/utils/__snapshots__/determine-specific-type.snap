// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (0) 1`] = `"type string ("'hello world'")"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (1) 1`] = `"type string ('hello worldðŸ‘‹ðŸŒŽðŸ‘‹ðŸŒŽðŸ‘‹ðŸŒŽðŸ‘‹...')"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (2) 1`] = `"type number (-0)"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (3) 1`] = `"type number (0)"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (4) 1`] = `"type bigint (13n)"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (5) 1`] = `"type number (3)"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (6) 1`] = `"type number (-Infinity)"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (7) 1`] = `"type number (NaN)"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (8) 1`] = `"type number (Infinity)"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (9) 1`] = `"type boolean (false)"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (10) 1`] = `"type symbol (Symbol(kIsNodeError))"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (11) 1`] = `"an instance of Date"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (12) 1`] = `"null"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (13) 1`] = `"type boolean (true)"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (14) 1`] = `"undefined"`;

exports[`unit:utils/determineSpecificType > should return specific type of \`value\` (15) 1`] = `"type function spy"`;
